src/main/host/syscall/futex.c

```diff
@@ -0,0 +1,164 @@
+#include "main/host/syscall/protected.h"
+#include "main/host/syscall_condition.h"
+#include "main/utility/utility.h"
+#include "support/logger/logger.h"
+
+// TODO: currently only supports uaddr from the same virtual address space (i.e., threads)
+// Support across different address spaces requires us to compute a unique id from the
+// hardware address (i.e., page table and offset). This is needed, e.g., when using
+// futexes across process boundaries.
+SysCallReturn syscallhandler_futex(SysCallHandler* sys, const SysCallArgs* args) {
+    utility_assert(sys && args);
+
+    PluginPtr uaddrptr = args->args[0].as_ptr; // int*
+    int futex_op = args->args[1].as_i64;
+    int val = args->args[2].as_i64;
+    PluginPtr timeoutptr = args->args[3].as_ptr; // const struct timespec*, or uint32_t
+    PluginPtr uaddr2ptr = args->args[4].as_ptr;  // int*
+    int val3 = args->args[5].as_i64;
+
+    const int possible_options = FUTEX_PRIVATE_FLAG | FUTEX_CLOCK_REALTIME;
+    int options = futex_op & possible_options;
+    int operation = futex_op & ~possible_options;
+
+    debug("futex called with addr=%p op=%i (operation=%i and options=%i) and val=%i",
+          (void*)uaddrptr.val, futex_op, operation, options, val);
+
+    // Futexes are 32 bits on all platforms
+    void* faddr = (uint32_t*)uaddrptr.val;
+    uint32_t expectedVal = (uint32_t)val;
+
+    // futex addr cannot be NULL
+    if (!faddr) {
+        debug("Futex addr cannot be NULL");
+        return (SysCallReturn){.state = SYSCALL_DONE, .retval.as_i64 = -EFAULT};
+    }
+
+    int result = 0;
+
+    switch (operation) {
+        case FUTEX_WAIT: {
+            debug("Handling FUTEX_WAIT operation %i", operation);
+
+            // This is a new wait operation on the futex for this thread.
+            // Check if a timeout was given in the syscall args.
+            const struct timespec* timeout;
+            if (timeoutptr.val) {
+                timeout =
+                    process_getReadablePtr(sys->process, sys->thread, timeoutptr, sizeof(*timeout));
+                // Bounds checking
+                if (!(timeout->tv_nsec >= 0 && timeout->tv_nsec <= 999999999)) {
+                    debug("A futex timeout was given, but the nanos value is out of range");
+                    result = -EINVAL;
+                    break;
+                }
+            } else {
+                timeout = NULL;
+            }
+
+            // TODO: does this load/compare need to be done atomically?
+            // `man 2 futex`: blocking via a futex is an atomic compare-and-block operation
+            const uint32_t* futexVal =
+                process_getReadablePtr(sys->process, sys->thread, uaddrptr, sizeof(uint32_t));
+
+            debug("Futex value is %" PRIu32 ", expected value is %" PRIu32, *futexVal, expectedVal);
+            if (*futexVal != expectedVal) {
+                debug("Futex values don't match, try again later");
+                result = -EAGAIN;
```

Comment:
Should this only be if `!_syscallhandler_wasBlocked(sys)`? e.g.:

* futex F has value X
* thread 1 waits on futex F with expected value X, causing it to get blocked
* thread 2 assigns Y to futex F and does a FUTEX_WAKE on it.

I believe in the current code thread 1 would be woken up but get -EAGAIN, whereas it should get woken up and get 0.

Reply:
I think you're right, good catch!